### PR TITLE
[export] add min/max ranges to _DimHints

### PR DESCRIPTION
**Pull request by pianpwk**
This adds min/max range versions of `Dim.AUTO` and `Dim.DYNAMIC`: `Dim.Auto(min=a, max=b) & Dim.Dynamic(min=a, max=b)`. This is nice for users who want the existing benefits of `Dim.AUTO/DYNAMIC`, but would like to encode properties like a maximum batch size into their program.

The intended behavior is for each placeholder symbol's range to be the **intersection** of the user-specified range, and the range from model tracing. If the latter range is more strict, then the user-specified range is effectively no-op, and a UserError is raised if there is no range overlap.

Example:
```
class Foo(torch.nn.Module):
    def forward(self, x, y):
        torch._check(x.shape[0] >= 4)
        torch._check(y.shape[0] <= 10)
        return x[2:] + y

inputs = (torch.randn(6), torch.randn(4))
shapes = {
    "x": {0: Dim.Dynamic(max=8)},
    "y": {0: Dim.AUTO(min=2)},
}
print(export(Foo(), inputs, dynamic_shapes=shapes))

ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, x: "f32[s1 + 2]", y: "f32[s1]"):
            ...
            return (add_2,)
Range constraints: {s1 + 2: VR[4, 8], s1: VR[2, 10]}
```

Implementation wise things are a bit sad because we have to support both strict + non-strict: we do a `ValueRanges.__and__` call in `_make_constraints`, which happens after produce_guards & the runtime asserts insertion pass. If we were to support non-strict only, we could refine these ranges with `torch._check` calls as part of a forward pre-hook to the original model. This would mean the min/max ranges are treated simply as additional relational guards, and go through the produce_guards process. This has benefits such as 1) eliminating any runtime assertions that become trivial, 2) better inference of ranges from relations & expressions, 3) having produce_guards + the Export solver as a single system for error handling. Unfortunately I couldn't figure out `torch._check` insertion for strict, and decided on this implementation to unify code paths.

Future follow ups:
- Figure out error messages & suggested fixes (?)
- Dynamic shapes serialization for these objects

### CODE REVIEW

File: torch/export/dynamic_shapes.py [hunk 1]
```diff
@@ -53,6 +53,26 @@ class _DimHint(Enum):
     DYNAMIC = auto()
 
 
+class _DimHint:
+    AUTO = _DimHintType.AUTO
+    STATIC = _DimHintType.STATIC
+    DYNAMIC = _DimHintType.DYNAMIC
+
+    def __init__(self, _type, _min=None, _max=None):
+        assert _type in (self.AUTO, self.STATIC, self.DYNAMIC)
```
Comment (avikchaudhuri) [inline torch/export/dynamic_shapes.py:62]: `STATIC`? Did you plan to add a `Static(min, max)` as well?

File: torch/export/dynamic_shapes.py [hunk 2]
```diff
@@ -53,6 +53,26 @@ class _DimHint(Enum):
     DYNAMIC = auto()
 
 
+class _DimHint:
+    AUTO = _DimHintType.AUTO
+    STATIC = _DimHintType.STATIC
+    DYNAMIC = _DimHintType.DYNAMIC
+
+    def __init__(self, _type, _min=None, _max=None):
```
Comment (avikchaudhuri) [inline torch/export/dynamic_shapes.py:61]: One nit I have with this design is that it's not clear that min and max are supposed to be "hints." E.g., hinting with `min=4` is compatible with the actual / effective `min` to be `8`, and values in this blind spot will result in runtime assertion errors. Do you have any ideas on how to address that?

File: torch/export/dynamic_shapes.py [hunk 3]
```diff
@@ -235,9 +255,11 @@ def Dim(name: str, *, min: Optional[int] = None, max: Optional[int] = None):
     return dim
 
 
-Dim.AUTO = _DimHint.AUTO  # type: ignore[attr-defined]
-Dim.STATIC = _DimHint.STATIC  # type: ignore[attr-defined]
-Dim.DYNAMIC = _DimHint.DYNAMIC  # type: ignore[attr-defined]
+Dim.AUTO = _DimHint(_DimHintType.AUTO)  # type: ignore[attr-defined]
```
Comment (Skylion007) [inline torch/export/dynamic_shapes.py:258]: Why not attach these as typing.ClassVariable so the type system understands?

File: torch/_export/serde/dynamic_shapes.py [hunk 4]
```diff
@@ -301,9 +301,11 @@ def deserialize_shape(
         if val is None or isinstance(val, int):
             return val
         elif val == "_DimHint.AUTO":
-            return _DimHint.AUTO
+            return Dim.AUTO  # type: ignore[attr-defined]
```
Comment (Skylion007) [inline torch/_export/serde/dynamic_shapes.py:304]: Why change this from before? It could infer the type properly before.
